programs/stake/src/stake_state.rs

```diff
@@ -614,14 +614,18 @@ pub fn delegate(
 }
 
 pub fn deactivate(
+    invoke_context: &InvokeContext,
     stake_account: &mut BorrowedAccount,
     clock: &Clock,
     signers: &HashSet<Pubkey>,
 ) -> Result<(), InstructionError> {
-    if let StakeState::Stake(meta, mut stake, stake_flags) = stake_account.get_state()? {
+    if let StakeState::Stake(meta, mut stake, mut stake_flags) = stake_account.get_state()? {
         meta.authorized.check(signers, StakeAuthorize::Staker)?;
-        stake.deactivate(clock.epoch)?;
-
+        let stake_history = invoke_context.get_sysvar_cache().get_stake_history()?;
+        stake.deactivate(clock.epoch, stake_flags, Some(&stake_history))?;
+        // After deactivation, need to clear DeactivationFlag.
+        // So that future activation and deactivation are not subject to the `MustFullyActivateBeforeDeactivationIsPermitted` restriction.
+        stake_flags.remove(StakeFlags::MUST_FULLY_ACTIVATE_BEFORE_DEACTIVATION_IS_PERMITTED);
```

Comment:
any reason not to pass the `InvokeContext` one level deeper into `Stake::deactivate()`? then we can dereplicode and do the `StakeHistory` query and flag removal in one place rather than every call site

Reply:
That will result in a dependency of `InvokeContext` which is in `solana-program` to sdk/programs/src/stake/state.rs. 

Currently that module only depends on class within `solana-sdk`. No foreign dependency outside its own library.

Reply:
ah. right... lame.  local helper, then?

Reply:
yeah. will do.

Reply:
Done.